### ISSUE TITLE
Multiple layer issues

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#329]
Here's some issues I identified while browsing the dl4j code. I'm not sure about some of these (and others are pretty minor), but thought it best to mention them.

-> BaseLayer.clone() (and subclass clone methods): Won't work in general.
(a) BaseLayer.clone() - Uses reflection; looks for a (NeuralNetConfiguration, INDArray, INDArray, INDArray) constructor (which doesn't exist in concrete subclasses such as BasePretrainNetwork, OutputLayer, RecursiveAutoEncoder, LSTM etc, which inherit the BaseLayer.clone() method).
(b) Even if a valid constructor did exist, the current clone implementation (in BaseLayer) won't work as intended (doesn't copy subclass fields for example in RecursiveAutoEncoder) though this of course could be done with reflection.

-> BaseLayer.Transpose(): Same problem as clone. Looks for same type of constructor that doesn't exist in subclasses.

-> BaseLayer.activate(): Line 297 if statement. Not sure if redundant or one branch of if statement is wrong.
```java
if(conf.getActivationFunction().equals("softmax"))
	return Nd4j.getExecutioner().execAndReturn(Nd4j.getOpFactory().createTransform(conf.getActivationFunction(), input().mmul(W).addiRowVector(b)));
else
	return Nd4j.getExecutioner().execAndReturn(Nd4j.getOpFactory().createTransform(conf.getActivationFunction(), input().mmul(W).addiRowVector(b)));
```

-> (Not sure, haven't checked the math) BaseLayer.backward(): has a variable called 'weightsPlusBias' that is actually just weights. Didn't seem correct to me at first glance.

-> OutputLayer.getWeightsGradient(): Calculated weight gradients for RMSE_XENT and SQUARED_LOSS are identical; RMSE_XENT is wrong, unless I'm mistaken?
```java
case RMSE_XENT:
	return input.transpose().mmul(pow(labels.sub(z),2));
case SQUARED_LOSS:
	return input.transpose().mmul(pow(labels.sub(z),2));
```
Also I'm assuming that the lack of a RECONSTRUCTION_CROSSENTROPY case in the same switch statement is ok?

-> DefaultLayerFactory.getParams(): Doesn't actually use the input Distribution object. Either DefaultLayerFactory.create() doesn't need to create (and pass to getParams) the distribution in the first place, or getParams() should be doing something with it. I'd guess the former -> minor clean-up issue.

**Comment by AlexDBlack** [comment]
Minor issue in RBM layer toString method (https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-core/src/main/java/org/deeplearning4j/nn/layers/feedforward/rbm/RBM.java); prints visible instead of hidden unit:
```java
return "RBM{" +
                ", visibleType=" + conf.getVisibleUnit() +
                ", hiddenType=" + conf.getVisibleUnit() +
                ", sigma=" + sigma +
                ", hiddenSigma=" + hiddenSigma +
                "} " + super.toString();
```

**Comment by agibsonccc** [comment]
I'm going to overhaul the layer toStrings(). If those inherit from base layer all of the toStrings() should be the same.

**Comment by agibsonccc** [comment]
So I fixed the RBM toStrings().

You're right about softmax there.

I fixed the layer clone. It now only assumes the conf (that was oversight on my part)

Re: BaseLayer activate() that was an artifact of OutputLayer. My bad there.

See:
https://github.com/deeplearning4j/deeplearning4j/blob/036eadbead0ab16efca4036df162aedaf5538419/deeplearning4j-core/src/main/java/org/deeplearning4j/nn/layers/OutputLayer.java#L344-344


Re: RECON. This is because it's specific to the sigmoid activation.


Re: backWard. Renamed this variable. That was a relic.

Closing this now. Thanks for the hard work. Feel free to take a look when you get the chance. Re open separate issues if there's anything else.
